### ISSUE TITLE
Windows Batch is horrible; PowerShell is nicer

### ISSUE THREAD

**Issue by nmigen-issue-migration** [nmigen/nmigen#252]
<a href="https://github.com/emilazy"><img src="https://avatars3.githubusercontent.com/u/18535642?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [emilazy](https://github.com/emilazy)**
_Sunday Oct 13, 2019 at 13:16 GMT_
_Originally opened as https://github.com/m-labs/nmigen/issues/252_

----

It would be nice to have less horrifying build scripts, especially as they increase in complexity due to #251 and similar.

**Comment by whitequark** [comment]
@emilazy Do you recall if there are any specific benefits from migrating to PowerShell, i.e. things we could not do before at all but could do after? I'm wary of a situation where I have to maintain *three* script generators because it turns out that some obscure but necessary configuration breaks PowerShell.

Batch files are awful, sure, but they do work, and migration infers a cost (both upstream and downstream) that should jutsify itself.

**Comment by whitequark** [comment]
> some obscure but necessary configuration breaks PowerShell

(I bet someone will try running batch files under wine cmd or something horrible like that...)

**Comment by programmerjake** [comment]
> (I bet someone will try running batch files under wine cmd or something horrible like that...)

:) report bugs to Wine, they will eventually get fixed (I fixed some myself). Also, if it works with Wine, it makes testing from Linux quite a bit easier.

**Comment by whitequark** [comment]
@programmerjake I assume batch files would work just fine under wine (I found wine cmd quite accurate). But I don't think wine has powershell by default. That would be one reason to stay on batch files.

**Comment by emilazy** [comment]
The Batch language probably has fundamental deficiencies with regards to things like filenames with spaces in them, but I don't know if any of them will really matter for nmigen.

On Linux or Wine you could probably use PowerShell Core, I guess? It's free software, which is more than can be said of the Batch interpreter.

That said, I think this was written when I was trying to move a bunch of logic from Python to build scripts, which ended up not happening, so you can close it if you'd like.

**Comment by whitequark** [comment]
> The Batch language probably has fundamental deficiencies with regards to things like filenames with spaces in them, but I don't know if any of them will really matter for nmigen.

nMigen explicitly forbids spaces (really, all non-alphanumeric characters) in filenames that it produces itself (e.g. the design name; the one that is `top` by default) for a simple reason: vendor toolchains don't handle them well. (People can still create such filenames explicitly, e.g. via `platform.add_file`.) Although I think handling such filenames does matter in principle, it seems like a lost cause given that a significant part of nMigen userbase will use it with third party software that I know will break on paths with spaces.

In other words, my judgement is that the benefits (improved path handling, language that's fit for its purpose) does not justify the drawbacks (churn in build scripts, added complications for wine users). I am open to reevaluating this in the future, but I think for now we should stick with batch files.
